pkg/auth_proxy/openstack.go

```diff
@@ -0,0 +1,154 @@
+package proxy
+
+import (
+	"fmt"
+	"github.com/shuaiming/mung/middlewares"
+	"net/http"
+	"time"
+
+	log "visualization-api/pkg/logging"
+)
+
+var (
+	// DefaultOpenStackGrafanaRolesMapping default roles mapping
+	DefaultOpenStackGrafanaRolesMapping = map[string]string{"admin": GrafanaRoleEditor,
+		"Member": GrafanaRoleReadOnlyEditor}
+)
+
+// OpenStackAuthHandler middleware for handling authentication via Keystone
+type OpenStackAuthHandler struct {
+	loginPage       []byte
+	grafanaStateTTL int
+	rolesMapping    map[string]string
+}
+
+// NewOpenStackAuthHandler returns OpenStackAuthHandler
+func NewOpenStackAuthHandler(loginPage []byte, grafanaStateTTL int, rolesMapping map[string]string) (*OpenStackAuthHandler, error) {
+	if rolesMapping == nil {
+		return &OpenStackAuthHandler{loginPage: loginPage,
+			grafanaStateTTL: grafanaStateTTL, rolesMapping: DefaultOpenStackGrafanaRolesMapping}, nil
+	}
+	return &OpenStackAuthHandler{loginPage: loginPage,
+		grafanaStateTTL: grafanaStateTTL, rolesMapping: rolesMapping}, nil
+}
+
+func (oh *OpenStackAuthHandler) ServeHTTP(rw http.ResponseWriter, r *http.Request, next http.HandlerFunc) {
```

Comment:
generally 'w' is used as response writer everywhere and also instead of 'oh' can you do it as 'handler'. this is just to maintain same pattern all over the code base.